test/annexB/built-ins/String/prototype/substr/start-and-length-as-numbers.js

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2022 André Bargull. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-string.prototype.substr
+description: >
+  Test String.prototype.substr with number inputs for start and length.
+---*/
+
+function ToIntegerOrInfinity(arg) {
+  assert.sameValue(typeof arg, "number");
+
+  return Number.isNaN(arg) ? 0 : Math.trunc(arg);
+}
+
+// Basic reference implementation. Expects all inputs have the correct type.
+function StringSubstr(string, start, length) {
+  // Steps 1-2.
+  assert.sameValue(typeof string, "string");
+
+  // Step 3.
+  let size = string.length;
+
+  // Step 4.
+  let intStart = ToIntegerOrInfinity(start);
+
+  // Steps 5-7.
+  if (intStart === -Infinity) {
+    intStart = 0;
+  } else if (intStart < 0) {
+    intStart = Math.max(size + intStart, 0);
+  } else {
+    intStart = Math.min(intStart, size)
+  }
+
+  // |0 <= intStart <= size| now holds.
+  assert(0 <= intStart && intStart <= size);
```

Comment:
I have a slight preference for turning these comments into assertion messages, e.g.
```js
assert(0 <= intStart && intStart <= size, "0 <= intStart <= size now holds");
```
That would've made it easier for me to debug when I was confused about why the implementation was different than in the spec. But it's a personal preference, I wouldn't block merging the PR on that.